### PR TITLE
Add tests for exceptions thrown from equals/hashcode methods in renderings.

### LINKED ISSUE #224: Exceptions thrown by RenderingAndSnapshot.equals are swallowed.
The `fail()` call in the test below is swallowed.

To see this exact failure, you'll have to revert the fix for #223 if that has landed. Or else replace the String rendering type with something that will throw from `equals`.

```kotlin
    lateinit var sink: Sink<String>

    val workflow = Workflow.stateful<Unit, String, Nothing, String>(
        initialState = { _, _ -> "unchanging state" },
        snapshot = { Snapshot.of {
          fail("Snapshot must not be resolved.")
        } },
        render = { _, state ->
          sink = actionSink.contraMap { action { this.state = it } }
          state
        }
    )
    val props = MutableStateFlow(Unit)
    val renderings = renderWorkflowIn(workflow, scope, props) {}

    renderings.value.rendering.let { rendered ->
      assertEquals("unchanging state", rendered)
      sink.send("unchanging state")
    }
```

### PR DESCRIPTION
**Pull request by zach-klippenstein**
These tests were written to catch situations like #224, which was actually fixed
in #225 by a PR to another issue. Nonetheless, these tests should help prevent
regressions.

Also makes other tests more strict by correctly calling `cleanupTestCoroutines()` on the default scope after the tests.

Closes #224 (doesn't really, it's obsolete, but close enough).

### CODE REVIEW

**Comment by zach-klippenstein** [comment]
@rjrjr Feel free to merge on approval.

File: workflow-runtime/src/test/java/com/squareup/workflow1/RenderWorkflowInTest.kt [hunk 1]
```diff
@@ -42,28 +46,43 @@ import kotlin.test.assertNotSame
...
+@OptIn(ExperimentalCoroutinesApi::class, FlowPreview::class)
 class RenderWorkflowInTest {
 
   // TestCoroutineScope doesn't actually create a Job, so isActive will always return true unless
   // explicitly give it a job.
-  private val scope = TestCoroutineScope(Job())
+
+  /**
+   * A [CoroutineScope] that will fail the test if it has uncaught exceptions after the test
+   * completes. Use this scope to test success cases.
```
Comment (rjrjr) [inline workflow-runtime/src/test/java/com/squareup/workflow1/RenderWorkflowInTest.kt:57]: Nice

**Review by rjrjr — APPROVED** [review]

File: workflow-runtime/src/test/java/com/squareup/workflow1/RenderWorkflowInTest.kt [hunk 2]
```diff
@@ -485,5 +507,85 @@ class RenderWorkflowInTest {
     )
   }
 
+  // https://github.com/square/workflow-kotlin/issues/224
+  @Test fun `exceptions from Snapshots don't fail runtime`() {
```
Comment (steve-the-edwards) [inline workflow-runtime/src/test/java/com/squareup/workflow1/RenderWorkflowInTest.kt:511]: Can you add to these test names 'don't fail runtime' shadows the fact that you are also testing that they *do fail* in Snapshot builder or equals when they should (i.e. don't get swallowed).

Comment (zach-klippenstein) [reply workflow-runtime/src/test/java/com/squareup/workflow1/RenderWorkflowInTest.kt:511]: Fair enough, although I think the better fix here is to not test the thing this test does not actually care about (stop testing that `throw` works).

**Pull request merged**